File: J/1306-jump-game-iii/main.go
package main

func canReach(arr []int, start int) bool {
	buf := make([]int, len(arr))
	stack := []int{start}

	for len(stack) > 0 {
		var t []int
		for i := 0; i < len(stack); i++ {
			if buf[stack[i]] == 1 {
				continue
			}
			buf[stack[i]] = 1
			if stack[i]+arr[stack[i]] < len(arr) && stack[i]+arr[stack[i]] > -1 {
				t = append(t, stack[i]+arr[stack[i]])
			}
			if stack[i]-arr[stack[i]] < len(arr) && stack[i]-arr[stack[i]] > -1 {
				t = append(t, stack[i]-arr[stack[i]])
			}
		}
		stack = t
	}

	for i := 0; i < len(arr); i++ {
		if arr[i] == 0 && buf[i] == 1 {
			return true
		}
	}
	return false
}

func main() {
	fmt.Println("a")
}

File: J/1306-jump-game-iii/main.rs
use std::collections::VecDeque;

impl Solution {
  pub fn can_reach(arr: Vec<i32>, start: i32) -> bool {
    let mut arr = arr;

    let mut q: VecDeque<i32> = VecDeque::new();
    q.push_back(start);

    while q.len() > 0 {
      let nxt = q.pop_front().unwrap() as usize;
      if arr[nxt] == 0 {
        return true;
      }

      if nxt as i32 - arr[nxt] >= 0 && arr[nxt - arr[nxt] as usize] != -1 {
        q.push_back(nxt as i32 - arr[nxt]);
      }
      if nxt as i32 + arr[nxt] < arr.len() as i32 && arr[nxt + arr[nxt] as usize] != -1 {
        q.push_back(nxt as i32 + arr[nxt]);
      }

      arr[nxt] = -1;
    }
    false
  }
}
